### PR TITLE
feat(contributing): add DevContainers for easier onboarding

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Previous behavior
We might have dependencies and tools (like linters) which are required to run Zebra and confirm everything is compliant or working even before submitting a PR, and not all developers might have all the required setup on their local machines.

## Expected behavior
Have a way for new developers or contributors to raise a Zebra development environment without having to change theirs, or add new tools for this purpose

## Solution
Use [Development Containers](https://containers.dev/) open specification for new contributors to easily setup a Zebra environment on their local machine, or use [GitHub Codespaces](https://docs.github.com/en/codespaces) by selecting the `devcontainer.json` configuration

Relates to #4464

## Review

Anyone can review this, but we might want to add documentation

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- Document this (where?)

### CODE REVIEW

**Comment by gustavovalverde** [comment]
Zebra's hardware minimum requirements might not be the minimum requirement for developing. But this is surely not my call. Anyways, this requirements does not allow people to use Codespaces 
image

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.245.2/containers/rust/.devcontainer/base.Dockerfile
+
+# [Choice] Debian OS version (use bullseye on local arm64/Apple Silicon): buster, bullseye
+ARG VARIANT="buster"
```
Comment (teor2345) [inline .devcontainer/Dockerfile]: bullseye is the later OS version, in general we prefer later versions to earlier ones.

This is also inconsistent with devcontainer.json, which uses bullseye.

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+FROM mcr.microsoft.com/vscode/devcontainers/rust:0-${VARIANT}
+
+# [Optional] Uncomment this section to install additional packages.
```
Comment (teor2345) [inline .devcontainer/Dockerfile]: These commented-out commands are also used below, so we can probably delete these commented-out duplicates.

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			"VARIANT": "bullseye"
+		}
+	},
+	"runArgs": [
+		"--cap-add=SYS_PTRACE",
+		"--security-opt",
+		"seccomp=unconfined"
+	],
+	"hostRequirements": {
+		"cpus": 2,
```
Comment (teor2345) [inline .devcontainer/devcontainer.json]: Building will take a long time with only 2 CPUs

```suggestion
		"cpus": 8,
```

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		"memory": "16gb",
```
Comment (teor2345) [inline .devcontainer/devcontainer.json:20]: If this is a problem we could try 8gb

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+				"panicbit.cargo",
+				"JScearcy.rust-doc-viewer",
+				"swellaby.vscode-rust-test-adapter",
+				"ms-vscode.test-adapter-converter"
+			]
+		}
+	},
+	// Use 'forwardPorts' to make a list of ports inside the container available locally.
+	"forwardPorts": [
+		18233,
```
Comment (teor2345) [inline .devcontainer/devcontainer.json:57]: Add the testnet RPC port, for consistency.

```suggestion
		18233,
		18232,
```

**Review by teor2345 — DISMISSED** [review]
This seems useful, but I'm concerned it is out of scope for our current DevOps work. I've been away, so I might have missed a team decision about this.

Is this something we want to spend time getting to work, or spend time supporting going forward?

Maybe @mpguerra can help with scope?

**Comment by mpguerra** [comment]
> This seems useful, but I'm concerned it is out of scope for our current DevOps work. I've been away, so I might have missed a team decision about this.
> 
> Is this something we want to spend time getting to work, or spend time supporting going forward?
> 
> Maybe @mpguerra can help with scope?

Yup, @gustavovalverde this is not currently a priority. We can review with low priority but I'd rather not spend too much more effort on this right now

**Comment by mpguerra** [comment]
> Yup, @gustavovalverde this is not currently a priority. We can review with low priority but I'd rather not spend too much more effort on this right now

I also just remembered that @gustavovalverde already mentioned this to me as an opportunistic update and my comment above still stands :)

**Comment by teor2345** [comment]
I'm pretty sure I don't have the right OS installed to do a proper review of this PR. (I don't have a GUI on my Linux machine.)

**Comment by conradoplg** [comment]
I tried testing this with VS Code by following https://code.visualstudio.com/docs/remote/containers but I get a error in rust-analyzer:

```
rust-analyzer failed to load workspace: Failed to read Cargo metadata from Cargo.toml file /workspaces/zebra/Cargo.toml, Some(Version { major: 1, minor: 64, patch: 0 }): Failed to run `"cargo" "metadata" "--format-version" "1" "--manifest-path" "/workspaces/zebra/Cargo.toml" "--filter-platform" "x86_64-unknown-linux-gnu"`: `cargo metadata` exited with an error:     Updating git repository `https://github.com/zcash/librustzcash.git`
 Downloading crates ...
  Downloaded clap v3.2.20
error: failed to open `/usr/local/cargo/registry/cache/github.com-1ecc6299db9ec823/clap-3.2.20.crate`

Caused by:
  Permission denied (os error 13)
```

**Comment by gustavovalverde** [comment]
This is for my own reference on solving the `cargo` issue: https://github.com/rust-lang/rust-analyzer/issues/3885#issuecomment-611076214

**Comment by mpguerra** [comment]
seems like this is not going to be a quick one to resolve so let's park it for now. We can ask our next joiner to be the guinea pig for this and let's close the PR in the meantime.

**Comment by gustavovalverde** [comment]
I'm moving it as a Draft PR and if I'm done with all the tickets of this Sprint and with spare time, I'll try to fix it. Otherwise I'll close it until we find a better time.

**Comment by teor2345** [comment]
> I'm moving it as a Draft PR and if I'm done with all the tickets of this Sprint and with spare time, I'll try to fix it. Otherwise I'll close it until we find a better time.

Sounds good!

Pili has asked me to check on sprint priorities for the next week, here are some DevOps tickets that could be a higher priority than this PR:
- #5091 (depends on PR #5481)
- #5044

**Comment by mpguerra** [comment]
I'm going to close this one for now :)

**Pull request closed without merging**